### ISSUE TITLE
Implicit initialization of omitted fields in conditionals

### ISSUE THREAD

**Issue by benharsh** [chapel-lang/chapel#8413]
### Summary of Problem
Consider the following class:
```chpl
class Tree {
  var left, right : Tree;

  proc init(depth : int) {
    if depth > 0 {
      left = new Tree(depth-1);
      right = new Tree(depth-1);
    }
  }
}
```

Is this a valid initializer?

In the [new initializer proposal](https://github.com/chapel-lang/chapel/issues/8283) this initializer will be in phase 1. In the past the compiler would require that all fields are initialized in all branches in phase 1, which would mean that the above code was invalid.

Alternatively, the compiler could default-init the omitted field in a compiler-generated else-branch. Would this be too surprising to users?

Related tests:
[test/studies/shootout/submitted/binarytrees.chpl](https://github.com/chapel-lang/chapel/tree/master/test/studies/shootout/submitted/binarytrees.chpl)

**Comment by bradcray** [comment]
@noakesmichael and I discussed a similar example that looked something like this (from memory):

```chapel
class C {
  var x, y, z: int;

  proc init() {
    if (test) {
      x = 1;
    } else {
      y = 2;
  }
  z = 3;
}
```

where one interpretation could that the compiler should add implicit initializations as follows based on what's initialized in, and after, each branch:

```chapel
class C {
  var x, y, z: int;

  proc init() {
    if (test) {
      x = 1;
      y = 0;
    } else {
      x = 0;
      y = 2;
    }
  }
  z = 3;
}
```

I'm generally in favor of this, as well as having an implicit `else` branch with default initializations get inserted in your case, though I can also see the fear of a slippery slope.  I don't view this as a necessary feature, but it seems like a nice one from a productive language that already supports default field initialization.  (And, of course, it'd be nice not to have to resubmit an updated version of this shootout if we didn't need to...  One of the reasons I wrote it without the `super.init()` was to try and make it forward portable...  didn't think about adding an `else` clause for similar reasons.
